swaywm/sway
# Implement swaynag

This is the first step towards #545.

This PR contains a clone of `i3-nagbar` with some added functionality. After some discussion on IRC, it was determined that we will do things a little different from i3 for showing errors. Instead of launching a terminal to show the errors, we will have a button to show detailed errors in a multi-line text box inside the expanded layer-surface.

The objective of this PR is to create `swaynagbar` in generic manner so it can be used with sway. This will not implement any actual use-cases, that will be left for future PRs.

**Left for follow-up PRs:**
- Show nagbar when there is config errors
- Add nagbar to the binding for `exit` in the default config (to match i3)
- (any other use cases that come up)

**Test Plan:**
- Show nagbar with a basic message
  - Make sure the dismiss button dismisses nagbar
- Show nagbar with a type of warning
- Show nagbar on a specific output
- Show nagbar on the bottom edge
- Show nagbar with an action button
  - Make sure the action is run when the button is clicked
- Show nagbar with a detailed message
  - Make sure nagbar expands and collapses
  - Make sure the text area can scroll with the mouse wheel and the buttons (requires long text)
  - Make sure a detailed message can be piped in via stdin
- Make sure config file works
  - Make sure default types (error and warning) can be overridden
  - Make sure custom types can be defined
  - Make sure command line args override config options
